pkg/reconciler/garbagecollector/garbagecollector_controller.go

```diff
@@ -0,0 +1,319 @@
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/discovery/cached/memory"
+	kubernetesclient "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/restmapper"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/component-base/metrics/prometheus/ratelimiter"
+	"k8s.io/klog/v2"
+	"k8s.io/kubernetes/pkg/controller/garbagecollector"
+
+	tenancyv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/tenancy/v1alpha1"
+	tenancyinformers "github.com/kcp-dev/kcp/pkg/client/informers/externalversions/tenancy/v1alpha1"
+	"github.com/kcp-dev/kcp/pkg/informer"
+	"github.com/kcp-dev/kcp/pkg/logging"
+	"github.com/kcp-dev/kcp/pkg/projection"
+)
+
+const (
+	ControllerName = "kcp-garbage-collector"
+)
+
+// Controller manages per-workspace garbage collector controllers.
+type Controller struct {
+	queue workqueue.RateLimitingInterface
+
+	dynamicDiscoverySharedInformerFactory *informer.DynamicDiscoverySharedInformerFactory
+	kubeClusterClient                     kubernetesclient.ClusterInterface
+	metadataClient                        kcpmetadata.ClusterInterface
+	informersStarted                      <-chan struct{}
+
+	workersPerLogicalCluster int
+
+	// lock guards the fields in this group
+	lock        sync.RWMutex
+	cancelFuncs map[logicalcluster.Name]func()
+
+	ignoredResources map[schema.GroupResource]struct{}
+
+	// For better testability
+	getClusterWorkspace func(key string) (*tenancyv1alpha1.ClusterWorkspace, error)
+	listCRDs            func() ([]*apiextensionsv1.CustomResourceDefinition, error)
+}
+
+// NewController creates a new Controller.
+func NewController(
+	clusterWorkspacesInformer tenancyinformers.ClusterWorkspaceInformer,
+	kubeClusterClient kubernetesclient.ClusterInterface,
+	metadataClient kcpmetadata.ClusterInterface,
+	dynamicDiscoverySharedInformerFactory *informer.DynamicDiscoverySharedInformerFactory,
+	crdInformer apiextensionsinformers.CustomResourceDefinitionInformer,
+	workersPerLogicalCluster int,
+	informersStarted <-chan struct{},
+) (*Controller, error) {
+	c := &Controller{
+		queue: workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), ControllerName),
+
+		dynamicDiscoverySharedInformerFactory: dynamicDiscoverySharedInformerFactory,
+		kubeClusterClient:                     kubeClusterClient,
+		metadataClient:                        metadataClient,
+		informersStarted:                      informersStarted,
+
+		workersPerLogicalCluster: workersPerLogicalCluster,
+
+		cancelFuncs: map[logicalcluster.Name]func(){},
+
+		ignoredResources: defaultIgnoredResources(),
+
+		getClusterWorkspace: func(key string) (*tenancyv1alpha1.ClusterWorkspace, error) {
+			return clusterWorkspacesInformer.Lister().Get(key)
+		},
+
+		listCRDs: func() ([]*apiextensionsv1.CustomResourceDefinition, error) {
```

Comment:
Doesn't this need a logical cluster to list in?

Reply:
It turns out this is not needed. Simply removed.